core/src/main/java/cucumber/runtime/RuntimeGlue.java

```diff
@@ -73,9 +74,26 @@ public StepDefinitionMatch stepDefinitionMatch(String featurePath, PickleStep st
 
     private List<StepDefinitionMatch> stepDefinitionMatches(String featurePath, PickleStep step) {
         List<StepDefinitionMatch> result = new ArrayList<StepDefinitionMatch>();
+
+        String stepText = step.getText();
+
+        CacheEntry cacheEntry = matchedStepDefinitionsCache.get(stepText);
+        if (cacheEntry != null) {
+            result.add(new StepDefinitionMatch(cacheEntry.arguments, cacheEntry.stepDefinition, featurePath, step, localizedXStreams));
+            return result;
+        }
+        boolean notFound = true;
         for (StepDefinition stepDefinition : stepDefinitionsByPattern.values()) {
             List<Argument> arguments = stepDefinition.matchedArguments(step);
             if (arguments != null) {
+                if (notFound) {
+                    matchedStepDefinitionsCache.put(stepText, new CacheEntry(stepDefinition, arguments));
+                    notFound = false;
+                }
+                else {
+                    //duplicate step - remove from cache
+                    matchedStepDefinitionsCache.remove(stepText);
+                }
```

Comment:
So rather then invalidating the cache entry when dealing with a duplicate step it might be better to store a list of cache entries. When upon fetching the list from the cache, the list contains multiple entries we can throw the `AmbiguousStepDefinitionsException`.